### ISSUE TITLE
Documentation - Algorithms

### ISSUE THREAD

**Issue by Lukas-Razz** [deeplearning4j/deeplearning4j#767]
It would be useful if you could write down basic information about:
-Optimalization Algorithms
-Loss Functions
-Updaters
-Unit types
That are implemented in dl4j. I know you could find them in papers, but i think some summarization would be useful so people using dl4j could now what they should/shouldn't use, how the work together any why. Some simple overview could be real help.

**Comment by tomthetrainer** [comment]
Yeah he is right, Note to myself, add this content to 
https://deeplearning4j.org/neuralnet-overview

**Comment by tomthetrainer** [comment]
build update here then create page. 

Text for Loss Functions. 

## DeepLearning4J supports the following Loss Functions. 

* MSE: Mean Squared Error: Linear Regression
* EXPLL: Exponential log likelihood: Poisson Regression
* XENT: Cross Entropy: Binary Classification
* MCXENT: Multiclass Cross Entropy
* RMSE_XENT: RMSE Cross Entropy
* SQUARED_LOSS: Squared Loss
* NEGATIVELOGLIKELIHOOD: Negative Log Likelihood

## Applyting Loss Functions

The Loss Function is applied when building your output Layer. 

```
layer(1, new OutputLayer.Builder(LossFunction.NEGATIVELOGLIKELIHOOD)
```

## Choosing the appropriate Loss Function

**Comment by tomthetrainer** [comment]
More text for update..

The JavaDoc for the Loss Function is part of ND4J javadoc and is available here. 
http://nd4j.org/doc/org/nd4j/linalg/api/ops/LossFunction.html

**Comment by tomthetrainer** [comment]
## Updaters

DL4J support the following Updaters

* ADADELTA 
* ADAGRAD 
* ADAM 
* CUSTOM 
* NESTEROVS 
* NONE 
* RMSPROP 
* SGD 

The JavaDoc for updaters is part of the DeepLearning4J JavaDoc and is available here
https://deeplearning4j.org/doc/org/deeplearning4j/nn/conf/Updater.html

## Optimization Algorithms

DeepLearning4J supports the following Optimization Algorithms. 

* CONJUGATE_GRADIENT 
* HESSIAN_FREE 
* LBFGS 
* LINE_GRADIENT_DESCENT 
* STOCHASTIC_GRADIENT_DESCENT

**Comment by tomthetrainer** [comment]
I do not know what the original request meant with regards to Unit types?

**Comment by AlexDBlack** [comment]
@tomthetrainer  Unit types: probably activation functions.
https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/activations/Activation.java#L11

Also don't forget that we now support custom layers, activation functions, loss functions.
(No custom updaters or optimization algorithms yet - but only researchers will tend to want to write those anyway - not practitioners. We have a CUSTOM enumeration for updaters, but I'd remove that - it isn't possible right now)

**Comment by tomthetrainer** [comment]
Note to Tom, OMG I am talking to myself. Add the ticket about gradient check to this, and add custom layers to this as Alex recommended

**Comment by tomthetrainer** [comment]
Closed

**Comment by dtitov** [comment]
@AlexDBlack, sorry for bringing this topic back after nearly a year, but a have a question about your statement: 

> No custom updaters or optimization algorithms yet - but only researchers will tend to want to write those anyway - not practitioners.

I'm actually representing researchers "cohort", so I would kinda need the possibility to implement custom optimization algorithms. Is it still the case that it's not possible or something changed since then?

**Comment by AlexDBlack** [comment]
@dtitov We support custom updaters used in conjunction with SGD (i.e., custom variants of Adam/momentum/etc). As for optimization algorithms (genetic algorithms, PSO, etc) it's not something we've added since then, we haven't needed it and there hasn't been much demand for it.
That said, in principle it shouldn't be too hard to add a few interfaces to allow supporting this. If that's what you need, feel free to open a new issue and we can discuss it further there.

For custom updaters, here's a unit test - https://github.com/deeplearning4j/deeplearning4j/blob/451dd76b50355358dc176f2b704e98c43423c5b8/deeplearning4j/deeplearning4j-core/src/test/java/org/deeplearning4j/nn/updater/custom/TestCustomUpdater.java
